drivers/sensor/vl53l1x/vl53l1.c

```diff
@@ -0,0 +1,485 @@
+		return -EINVAL;
+	}
+
+	/* Map val to pUserROi */
+	VL53L1_UserRoi_t pUserROi = {
+		.TopLeftX = val->val1 % 16,
+		.TopLeftY = (uint8_t)(val->val1 / 16),
+		.BotRightX = val->val2 % 16,
+		.BotRightY = (uint8_t)(val->val2 / 16),
+	};
+
+	ret = VL53L1_SetUserROI(&drv_data->vl53l1x, &pUserROi);
+	if (ret != VL53L1_ERROR_NONE) {
+		LOG_ERR("[%s] VL53L1_SetUserROI return error (%d)", dev->name, ret);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int vl53l1x_get_mode(const struct device *dev,
+		struct sensor_value *val)
+{
+	struct vl53l1x_data *drv_data = dev->data;
+	VL53L1_DistanceModes mode;
+	VL53L1_Error ret;
+
+	ret = VL53L1_GetDistanceMode(&drv_data->vl53l1x, &mode);
+	if (ret != VL53L1_ERROR_NONE) {
+		LOG_ERR("[%s] VL53L1_GetDistanceMode return error (%d)", dev->name, ret);
+		return -EINVAL;
+	}
+
+	/* Mapping is 1:1 with the API */
+	val->val1 = (int32_t)mode;
+	val->val2 = 0;
+	return 0;
+}
+
+static int vl53l1x_get_roi(const struct device *dev,
+		struct sensor_value *val)
+{
+	struct vl53l1x_data *drv_data = dev->data;
+	VL53L1_Error ret;
+	VL53L1_UserRoi_t pUserROi;
+
+	ret = VL53L1_GetUserROI(&drv_data->vl53l1x, &pUserROi);
+	if (ret != VL53L1_ERROR_NONE) {
+		LOG_ERR("[%s] VL53L1_GetUserROI return error (%d)", dev->name, ret);
+		return -EINVAL;
+	}
+
+	/* Map pUserROi to val */
+	val->val1 = (int32_t)((16 * pUserROi.TopLeftY) + pUserROi.TopLeftX);
+	val->val2 = (int32_t)((16 * pUserROi.BotRightY) + pUserROi.BotRightX);
+	return 0;
+}
+
+static int vl53l1x_sample_fetch(const struct device *dev,
+		enum sensor_channel chan)
+{
+	struct vl53l1x_data *drv_data = dev->data;
+	VL53L1_Error ret;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_DISTANCE);
+
+	/* Will immediately stop current measurement */
+	ret = VL53L1_StopMeasurement(&drv_data->vl53l1x);
+	if (ret != VL53L1_ERROR_NONE) {
+		LOG_ERR("VL53L1_StopMeasurement return error (%d)", ret);
+		return -EINVAL;
+	}
+
+	if (IS_ENABLED(CONFIG_VL53L1X_INTERRUPT_MODE)) {
+		vl53l1x_enable_interrupt(drv_data->dev, true);
+	}
+
+	ret = VL53L1_StartMeasurement(&drv_data->vl53l1x);
+	if (ret != VL53L1_ERROR_NONE) {
+		LOG_ERR("[%s] VL53L1_StartMeasurement return error (%d)", dev->name, ret);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int vl53l1x_channel_get(const struct device *dev,
+		enum sensor_channel chan,
+		struct sensor_value *val)
+{
+	struct vl53l1x_data *drv_data = dev->data;
+	VL53L1_Error ret;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_DISTANCE);
+
+	/* Calling VL53L1_WaitMeasurementDataReady regardless of using interrupt or
+	 * polling method ensures user does not have to consider the time between
+	 * calling fetch and get.
+	 */
+	ret = VL53L1_WaitMeasurementDataReady(&drv_data->vl53l1x);
+	if (ret != VL53L1_ERROR_NONE) {
+		LOG_ERR("[%s] VL53L1_WaitMeasurementDataReady return error (%d)", dev->name, ret);
+		return -EINVAL;
+	}
+
+	if (IS_ENABLED(CONFIG_VL53L1X_INTERRUPT_MODE) == 0) {
+		/* Using driver poling mode */
+		vl53l1x_read_sensor(drv_data);
```

Comment:
It would nice to the user to pass any errors from vl53l1x_read_sensor() on here.  vl53l1x_channel_get() should return a failure code instead of returning success when there is an error.  vl53l1x_channel_get() would also have to not ignore errors.

Reply:
Great catch. Thanks. I went over the return codes too and I feel much better now.